File: examples/tv-app/android/App/platform-app/src/main/java/com/matter/tv/server/MainActivity.java
package com.matter.tv.server;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.provider.Settings;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import com.google.android.material.bottomnavigation.BottomNavigationView;
import com.matter.tv.server.fragments.ContentAppFragment;
import com.matter.tv.server.fragments.QrCodeFragment;
import com.matter.tv.server.fragments.TerminalFragment;
import java.util.LinkedHashMap;

public class MainActivity extends AppCompatActivity {

  private LinkedHashMap<String, String> packages = new LinkedHashMap<>();

  private BottomNavigationView.OnNavigationItemSelectedListener navListener =
      item -> {
        Fragment selectedFragment = null;
        switch (item.getItemId()) {
          case R.id.content_app:
            selectedFragment = ContentAppFragment.newInstance();
            break;
          case R.id.qr_code:
            selectedFragment = QrCodeFragment.newInstance();
            break;
          case R.id.terminal:
            selectedFragment = TerminalFragment.newInstance();
            break;
        }

        getSupportFragmentManager()
            .beginTransaction()
            .replace(R.id.fragment_container_view, selectedFragment)
            .commit();
        return true;
      };

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);

    BottomNavigationView bottomNavigationView = findViewById(R.id.bottom_navigation);
    bottomNavigationView.setOnItemSelectedListener(navListener);

    getSupportFragmentManager()
        .beginTransaction()
        .replace(R.id.fragment_container_view, new QrCodeFragment())
        .commit();
    checkOverlayPermission();
  }

  private void checkOverlayPermission() {
    if (!Settings.canDrawOverlays(this)) {
      AlertDialog.Builder builder = new AlertDialog.Builder(this);
      builder
          .setMessage("Allow permission to display over other apps")
          .setTitle("Request overlay permission")
          .setPositiveButton(
              "Ok",
              (dialog, which) -> {
                dialog.dismiss();
                Intent intent = new Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION);
                intent.setData(Uri.parse("package:" + getPackageName()));
                startActivity(intent);
              })
          .create()
          .show();
    }
  }
}

File: examples/virtual-device-app/android/App/app/src/main/java/com/matter/virtual/device/app/MainActivity.kt
package com.matter.virtual.device.app

import android.Manifest
import android.annotation.SuppressLint
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.os.PowerManager
import android.provider.Settings
import android.view.Gravity
import android.view.MenuItem
import android.view.View
import androidx.activity.result.contract.ActivityResultContracts
import androidx.activity.viewModels
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import com.matter.virtual.device.app.core.common.EventObserver
import com.matter.virtual.device.app.core.ui.SharedViewModel
import com.matter.virtual.device.app.core.ui.UiState
import com.matter.virtual.device.app.databinding.ActivityMainBinding
import dagger.hilt.android.AndroidEntryPoint
import timber.log.Timber

@AndroidEntryPoint
class MainActivity : AppCompatActivity() {

  private lateinit var binding: ActivityMainBinding
  private val viewModel by viewModels<SharedViewModel>()

  private val permissions =
    arrayOf(
      Manifest.permission.BLUETOOTH_SCAN,
      Manifest.permission.BLUETOOTH_ADVERTISE,
      Manifest.permission.BLUETOOTH_CONNECT,
      Manifest.permission.ACCESS_FINE_LOCATION
    )

  private val requestMultiplePermissions =
    registerForActivityResult(ActivityResultContracts.RequestMultiplePermissions()) { permissions ->
      permissions.entries.forEach { Timber.d("${it.key}:${it.value}") }
    }

  @SuppressLint("BatteryLife")
  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    Timber.d("onCreate()")

    val isPermissionsGranted =
      permissions.all {
        ContextCompat.checkSelfPermission(this, it) == PackageManager.PERMISSION_GRANTED
      }

    if (!isPermissionsGranted) {
      requestMultiplePermissions.launch(permissions)
    }

    val powerManager = applicationContext.getSystemService(POWER_SERVICE) as PowerManager?
    powerManager?.let { manager ->
      if (!manager.isIgnoringBatteryOptimizations(packageName)) {
        Timber.d("not in battery optimization whitelist")
        val intent =
          Intent().apply {
            action = Settings.ACTION_REQUEST_IGNORE_BATTERY_OPTIMIZATIONS
            data = Uri.parse("package:$packageName")
          }
        startActivity(intent)
      }
    }

    binding = ActivityMainBinding.inflate(layoutInflater)
    setContentView(binding.root)

    viewModel.uiState.observe(
      this,
      EventObserver { uiState ->
        when (uiState) {
          UiState.Waiting -> {
            binding.progress.visibility = View.VISIBLE
          }
          UiState.Exit -> {
            binding.progress.visibility = View.GONE
            finishAffinity()
          }
          is UiState.Reset -> {
            showFactoryResetPopup(getString(uiState.messageResId), uiState.isCancelable)
          }
          else -> {}
        }
      }
    )
  }

  override fun onDestroy() {
    super.onDestroy()
    Timber.d("onDestroy()")
  }

  override fun onOptionsItemSelected(item: MenuItem): Boolean {
    when (item.itemId) {
      android.R.id.home -> {
        onBackPressed()
        return true
      }
    }

    return super.onOptionsItemSelected(item)
  }

  override fun onRequestPermissionsResult(
    requestCode: Int,
    permissions: Array<out String>,
    grantResults: IntArray
  ) {
    Timber.d("RequestCode:$requestCode")
    super.onRequestPermissionsResult(requestCode, permissions, grantResults)
  }

  private fun showFactoryResetPopup(message: String, isCancelable: Boolean) {
    val builder =
      AlertDialog.Builder(this)
        .setTitle("Factory Reset")
        .setMessage(message)
        .setPositiveButton("Ok") { dialog, _ ->
          Timber.d("Ok")
          dialog.dismiss()
          viewModel.resetMatterAppServer()
        }
        .setCancelable(false)

    if (isCancelable) {
      builder.setNegativeButton("Cancel") { dialog, _ ->
        Timber.d("Cancel")
        dialog.dismiss()
      }
    }

    val dialog = builder.create()
    dialog.window?.setGravity(Gravity.BOTTOM)
    dialog.show()
  }
}
